### PR TITLE
[feat] Add better support for predict + ddp 2/3

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Here is PR 1/2: https://github.com/PyTorchLightning/pytorch-lightning/pull/7141
This PR adds better support for predict.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/overrides/distributed.py [hunk 1]
```diff
@@ -75,3 +76,70 @@ def prepare_for_backward(model: DistributedDataParallel, output: Any):
...
+
+
+class IndexBatchSamplerWrapper:
+    """This class is used to wrap a :class:`torch.utils.data.BatchSampler` and capture its indices."""
+
+    def __init__(self, sampler: BatchSampler) -> None:
+        self.batch_sampler = sampler
+        self.batch_indices: Optional[List[int]] = None
+
+    def __iter__(self) -> Iterator[List[int]]:
```
Comment (kaushikb11) [inline pytorch_lightning/overrides/distributed.py:130]: The return type is Iterator?

File: pytorch_lightning/core/datamodule.py [hunk 2]
```diff
@@ -26,7 +26,7 @@
 
 class _DataModuleWrapper(type):
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, **kwargs) -> None:
```
Comment (ethanwharris) [inline pytorch_lightning/core/datamodule.py:29]:
```suggestion
    def __init__(self, *args: Any, **kwargs: Any) -> None:
```

File: pytorch_lightning/core/lightning.py [hunk 3]
```diff
@@ -74,7 +74,7 @@ class LightningModule(
         "model_size",
     ] + DeviceDtypeModuleMixin.__jit_unused_properties__
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, **kwargs) -> None:
```
Comment (ethanwharris) [inline pytorch_lightning/core/lightning.py:77]:
```suggestion
    def __init__(self, *args: Any, **kwargs: Any) -> None:
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 4]
```diff
@@ -73,6 +73,11 @@ def model_to_device(self) -> None:
     def is_global_zero(self) -> bool:
         """Whether the current process is the rank zero process not only on the local node, but for all nodes."""
 
+    @property
+    @abstractmethod
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:77]: Maybe don't have it as `abstractmethod`, and keep the default as False

File: pytorch_lightning/overrides/distributed.py [hunk 5]
```diff
@@ -75,3 +76,70 @@ def prepare_for_backward(model: DistributedDataParallel, output: Any):
...
+
+class UnrepeatedDistributedSampler(DistributedSampler):
+    """
+    A fork of the pytorch DistributedSampler that doesn't repeat data, instead
+    allowing the number of batches per process to be off-by-one from each other.
+    This makes this sampler usable for validation (it's deterministic and
+    doesn't require shuffling). It is potentially unsafe to use this sampler for
+    training, because during training the DistributedDataParallel syncs buffers
+    on each forward pass, so it could freeze if one of the processes runs one
+    fewer batch. During validation, buffers are only synced on the first batch,
```
Comment (ethanwharris) [inline pytorch_lightning/overrides/distributed.py:89]: We only use for predicting but the comment says it can also be for validation?

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley: small comments

File: pytorch_lightning/core/datamodule.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/core/datamodule.py:29]: Done !

File: pytorch_lightning/overrides/distributed.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:89]: Good catch !

File: pytorch_lightning/plugins/training_type/training_type_plugin.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/training_type_plugin.py:77]: It is safer to make it abstract.

File: pytorch_lightning/overrides/distributed.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:130]: Yes, it yield a list.

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/predict_loop.py [hunk 6]
```diff
@@ -24,6 +27,28 @@ def __init__(self, trainer):
...
+        self.epoch_batch_indices: Optional[List[List[int]]] = None
+        self._return_predictions = not trainer.training_type_plugin.use_spawn
+        self._previous_grad_status: Optional[bool] = None
+
+    @property
+    def return_predictions(self) -> bool:
+        return self._return_predictions
+
+    @return_predictions.setter
+    def return_predictions(self, return_predictions: Optional[bool] = None) -> None:
```
Comment (Borda) [inline pytorch_lightning/trainer/predict_loop.py:40]: what is the case of passing None?

File: pytorch_lightning/trainer/predict_loop.py [hunk 7]
```diff
@@ -24,6 +27,28 @@ def __init__(self, trainer):
...
+        if return_predictions and self.trainer.training_type_plugin.use_spawn:
+            raise MisconfigurationException(
+                "`return_predictions` should be set to `False` when using spawn accelerators. "
+                f"Found {return_predictions}"
+            )
+        if return_predictions is not None:
+            self._return_predictions = return_predictions
+
+    @property
+    def should_store_predictions(self) -> bool:
```
Comment (Borda) [inline pytorch_lightning/trainer/predict_loop.py:50]: why extra property with direct mapping to another property?

File: pytorch_lightning/trainer/predict_loop.py [hunk 8]
```diff
@@ -82,18 +111,43 @@ def predict_step(self, batch, batch_idx, dataloader_idx):
 
         self.trainer.call_hook("on_predict_batch_end", predictions, batch, batch_idx, dataloader_idx)
 
-        self._predictions[dataloader_idx].append(predictions)
+        if self.should_store_predictions:
```
Comment (Borda) [inline pytorch_lightning/trainer/predict_loop.py:114]:
```suggestion
        if self.return_predictions:
```

File: pytorch_lightning/trainer/predict_loop.py [hunk 9]
```diff
@@ -82,18 +111,43 @@ def predict_step(self, batch, batch_idx, dataloader_idx):
...
         self.trainer.call_hook("on_predict_epoch_end", results)
 
-        def _convert_to_numpy(v):
-            return v.cpu().numpy()
+        if self.return_predictions:
+            return results[0] if self.num_dataloaders == 1 else results
 
-        results = apply_to_collection(results, torch.Tensor, _convert_to_numpy)
+    def on_predict_end(self):
+        # clean memory
```
Comment (Borda) [inline pytorch_lightning/trainer/predict_loop.py:146]: mind add a comment about where the predictions go?

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -1049,6 +1041,10 @@ def predict(
 
             datamodule: A instance of :class:`LightningDataModule`.
 
+            return_predictions: Whether to return predictions.
+                By default, it will be set to ``True``
+                except when an accelerator that spawns processes is used (not supported).
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1046]:
```suggestion
                By default, it except when an accelerator that spawns processes is used (not supported).
```

File: pytorch_lightning/utilities/xla_device.py [hunk 11]
```diff
@@ -87,14 +86,6 @@ def tpu_device_exists() -> bool:
...
-        if os.getenv("PL_TPU_AVAILABLE", '0') == "1":
-            XLADeviceUtils._TPU_AVAILABLE = True
-
         if XLADeviceUtils.xla_available() and not XLADeviceUtils._TPU_AVAILABLE:
-
             XLADeviceUtils._TPU_AVAILABLE = XLADeviceUtils._is_device_tpu()
-
-            if XLADeviceUtils._TPU_AVAILABLE:
-                os.environ["PL_TPU_AVAILABLE"] = '1'
-
```
Comment (Borda) [inline pytorch_lightning/utilities/xla_device.py:99]: shall this be in separate PR?

File: tests/loggers/test_all.py [hunk 12]
```diff
@@ -102,7 +102,7 @@ def test_epoch_end(self, outputs) -> None:
 
     class StoreHistoryLogger(logger_class):
 
-        def __init__(self, *args, **kwargs):
+        def __init__(self, *args, **kwargs) -> None:
```
Comment (Borda) [inline tests/loggers/test_all.py:105]: I do not think that init are annotated with None cc: @carmocca 
```suggestion
        def __init__(self, *args, **kwargs):
```

File: tests/loggers/test_base.py [hunk 13]
```diff
@@ -170,7 +170,7 @@ def test_adding_step_key(tmpdir):
 
     class CustomTensorBoardLogger(TensorBoardLogger):
 
-        def __init__(self, *args, **kwargs):
+        def __init__(self, *args, **kwargs) -> None:
```
Comment (Borda) [inline tests/loggers/test_base.py:173]: leave this changes to separate PR

File: pytorch_lightning/utilities/imports.py [hunk 14]
```diff
@@ -92,4 +93,4 @@ def _compare_version(package: str, op, version) -> bool:
 
 from pytorch_lightning.utilities.xla_device import XLADeviceUtils  # noqa: E402
 
-_TPU_AVAILABLE = XLADeviceUtils.tpu_device_exists()
+_TPU_AVAILABLE = XLADeviceUtils.tpu_device_exists() or int(os.getenv("PL_TPU_AVAILABLE", 0))
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/imports.py:96]: Why is this required in this PR?

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/utilities/imports.py (hunk 14 above)
Comment (tchaton) [reply pytorch_lightning/utilities/imports.py:96]: Trying to mock TPU, but this doesn't work.

File: pytorch_lightning/overrides/distributed.py [hunk 15]
```diff
@@ -75,3 +76,70 @@ def prepare_for_backward(model: DistributedDataParallel, output: Any):
             model.reducer.prepare_for_backward([])
     else:
         model.require_forward_param_sync = False
+
+
+class UnrepeatedDistributedSampler(DistributedSampler):
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:81]: It means the data won't be repeated.

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:81]: I would prefer to keep `Unrepeated` in this case.

File: pytorch_lightning/overrides/distributed.py [hunk 16]
```diff
@@ -75,3 +76,70 @@ def prepare_for_backward(model: DistributedDataParallel, output: Any):
...
+        assert self.num_samples >= 1 or self.total_size == 0
+
+    def __iter__(self) -> Iterator[List[int]]:
+        if self.shuffle:
+            # deterministically shuffle based on epoch
+            g = torch.Generator()
+            g.manual_seed(self.epoch)
+            indices = torch.randperm(len(self.dataset), generator=g).tolist()
+        else:
+            indices = list(range(len(self.dataset)))
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:112]: I would say yes as we check the size.

File: pytorch_lightning/overrides/distributed.py [hunk 17]
```diff
@@ -75,3 +76,70 @@ def prepare_for_backward(model: DistributedDataParallel, output: Any):
...
+
+        return iter(indices)
+
+
+class IndexBatchSamplerWrapper:
+    """This class is used to wrap a :class:`torch.utils.data.BatchSampler` and capture its indices."""
+
+    def __init__(self, sampler: BatchSampler) -> None:
+        self.batch_sampler = sampler
+        self.batch_indices: Optional[List[int]] = None
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:128]: Yes, because we access it from outside and it is slightly easier to understand what indices means.

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:128]: Yes, let's make the sampler private.

File: pytorch_lightning/overrides/distributed.py [hunk 18]
```diff
@@ -75,3 +76,70 @@ def prepare_for_backward(model: DistributedDataParallel, output: Any):
...
+
+    def __iter__(self) -> Iterator[List[int]]:
+        for batch in self.batch_sampler:
+            self.batch_indices = batch
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/overrides/distributed.py:132]: Same as before, let's keep it public.

File: pytorch_lightning/trainer/predict_loop.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:40]: None is the default. It means the training type plugin will set its own default.
`return_predictions = not training_type_plugin.use_spawn`

File: pytorch_lightning/trainer/predict_loop.py (hunk 7 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:50]: It was from the PR 3/3. Removed !

File: pytorch_lightning/trainer/predict_loop.py (hunk 8 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:114]: Done !

File: pytorch_lightning/trainer/predict_loop.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:146]: Done !

File: pytorch_lightning/trainer/trainer.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1046]: Done !

File: pytorch_lightning/utilities/xla_device.py (hunk 11 above)
Comment (tchaton) [reply pytorch_lightning/utilities/xla_device.py:99]: Put it back !

File: tests/loggers/test_all.py (hunk 12 above)
Comment (tchaton) [reply tests/loggers/test_all.py:105]: removed !

File: pytorch_lightning/utilities/imports.py (hunk 14 above)
Comment (Borda) [reply pytorch_lightning/utilities/imports.py]: then pls rever it

File: pytorch_lightning/overrides/distributed.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/overrides/distributed.py:81]: won't be repeated in what sense?

File: pytorch_lightning/overrides/distributed.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (Borda) [reply pytorch_lightning/overrides/distributed.py:132]: above https://github.com/PyTorchLightning/pytorch-lightning/pull/7215#discussion_r620547440 you say private so what you mean?

File: tests/loggers/test_all.py (hunk 12 above)
Comment (Borda) [reply tests/loggers/test_all.py]: everywhere? bc it was on about 10 other places in this PR

File: tests/overrides/test_distributed.py [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def test_unrepeated_distributed_sampler(tmpdir):
+    """Test each rank will receive a different number of elements."""
+
+    world_size = 4
+    samplers = []
+    dataset = range(103)
+    for rank in range(world_size):
+        samplers.append(UnrepeatedDistributedSampler(dataset, rank=rank, num_replicas=world_size, shuffle=False))
```
Comment (awaelchli) [inline tests/overrides/test_distributed.py:27]: the shuffle=True branch has no coverage

File: pytorch_lightning/trainer/data_loading.py [hunk 20]
```diff
@@ -133,20 +137,23 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
...
 
         return dataloader
 
     @staticmethod
-    def _resolve_batch_sampler(dl_args, dataloader, sampler):
+    def _resolve_batch_sampler(dl_args, dataloader, sampler, mode: Optional[RunningStage] = None) -> Dict[str, Any]:
         batch_sampler = getattr(dataloader, "batch_sampler")
-        if batch_sampler is not None and type(batch_sampler) is not BatchSampler:
+        is_predicting = mode == RunningStage.PREDICTING
+        if (batch_sampler is not None and type(batch_sampler) is not BatchSampler) or is_predicting:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:149]: We are checking the type, not the instance. I find it cleaner this way and more explicit.

File: pytorch_lightning/trainer/data_loading.py [hunk 21]
```diff
@@ -133,20 +137,23 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
...
             batch_sampler = type(batch_sampler)(
                 sampler,
                 batch_size=batch_sampler.batch_size,
-                drop_last=batch_sampler.drop_last,
+                drop_last=False if is_predicting else batch_sampler.drop_last,
             )
+            if is_predicting:
+                batch_sampler = IndexBatchSamplerWrapper(batch_sampler)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:156]: Done !

File: tests/overrides/test_distributed.py [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+
+def test_index_batch_sampler(tmpdir):
+    """Test `IndexBatchSampler` properly extracts indices."""
+    dataset = range(15)
+    sampler = SequentialSampler(dataset)
+    batch_sampler = BatchSampler(sampler, 3, False)
+    index_batch_sampler = IndexBatchSamplerWrapper(batch_sampler)
+
+    for batch in index_batch_sampler:
+        assert index_batch_sampler.batch_indices == batch
```
Comment (awaelchli) [inline tests/overrides/test_distributed.py:49]:
```suggestion
    assert batch_sampler.batch_size == index_batch_sampler.batch_size
    assert batch_sampler.drop_last == index_batch_sampler.drop_last
    assert batch_sampler.sampler is sampler
    
    for batch in index_batch_sampler:
        assert index_batch_sampler.batch_indices == batch
```

File: pytorch_lightning/trainer/predict_loop.py [hunk 23]
```diff
@@ -54,7 +75,8 @@ def setup(self, model, max_batches, dataloaders):
 
         self.max_batches = max_batches
         self.num_dataloaders = self._get_num_dataloaders(dataloaders)
-        self._predictions = [[] for _ in range(self.num_dataloaders)]
+        self.predictions = [[] for _ in range(self.num_dataloaders)]
+        self.epoch_batch_indices = [[] for _ in range(self.num_dataloaders)]
 
     def _get_num_dataloaders(self, dataloaders):
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:81]: Done !

File: pytorch_lightning/plugins/training_type/parallel.py [hunk 24]
```diff
@@ -47,6 +47,10 @@ def root_device(self):
     def on_gpu(self):
         return self.root_device.type == "cuda" and torch.cuda.is_available()
 
+    @property
+    def use_spawn(self) -> bool:
+        return False
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/parallel.py:52]: I would avoid this. Can we do isinstance check?

File: pytorch_lightning/trainer/data_loading.py [hunk 25]
```diff
@@ -133,20 +137,23 @@ def auto_add_sampler(self, dataloader: DataLoader, shuffle: bool) -> DataLoader:
...
+    def _resolve_batch_sampler(dl_args, dataloader, sampler, mode: Optional[RunningStage] = None) -> Dict[str, Any]:
         batch_sampler = getattr(dataloader, "batch_sampler")
-        if batch_sampler is not None and type(batch_sampler) is not BatchSampler:
+        is_predicting = mode == RunningStage.PREDICTING
+        if (batch_sampler is not None and type(batch_sampler) is not BatchSampler) or is_predicting:
             batch_sampler = type(batch_sampler)(
                 sampler,
                 batch_size=batch_sampler.batch_size,
-                drop_last=batch_sampler.drop_last,
+                drop_last=False if is_predicting else batch_sampler.drop_last,
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py:154]:
```suggestion
                drop_last=(False if is_predicting else batch_sampler.drop_last),
```
so the user returning a dataloader with `drop_last=True for` for prediction will be ignored?

File: pytorch_lightning/trainer/predict_loop.py [hunk 26]
```diff
@@ -82,18 +107,43 @@ def predict_step(self, batch, batch_idx, dataloader_idx):
...
+            return results[0] if self.num_dataloaders == 1 else results
 
-        results = apply_to_collection(results, torch.Tensor, _convert_to_numpy)
+    def on_predict_end(self):
+        # clean memory. the predictions are extracted in `on_predict_epoch_end`.
+        self.predictions = None
+        self.batch_indices = None
 
-        return results[0] if len(results) == 1 else results
+        # enable eval mode + no grads
```
Comment (awaelchli) [inline pytorch_lightning/trainer/predict_loop.py:147]: comment is not accurate

File: pytorch_lightning/trainer/predict_loop.py [hunk 27]
```diff
@@ -82,18 +107,43 @@ def predict_step(self, batch, batch_idx, dataloader_idx):
...
         self.trainer.call_hook("on_predict_epoch_end", results)
 
-        def _convert_to_numpy(v):
-            return v.cpu().numpy()
+        if self.return_predictions:
+            return results[0] if self.num_dataloaders == 1 else results
 
-        results = apply_to_collection(results, torch.Tensor, _convert_to_numpy)
+    def on_predict_end(self):
+        # clean memory. the predictions are extracted in `on_predict_epoch_end`.
```
Comment (awaelchli) [inline pytorch_lightning/trainer/predict_loop.py:143]:
```suggestion
        # clear memory. the predictions are extracted in `on_predict_epoch_end`.
```

File: pytorch_lightning/trainer/trainer.py [hunk 28]
```diff
@@ -1049,6 +1041,9 @@ def predict(
 
             datamodule: A instance of :class:`LightningDataModule`.
 
+            return_predictions: Whether to return predictions.
+                By default, it except when an accelerator that spawns processes is used (not supported).
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1045]:
```suggestion
                ``True`` by default except when an accelerator that spawns processes is used (not supported).
```

File: pytorch_lightning/trainer/predict_loop.py [hunk 29]
```diff
@@ -24,6 +27,24 @@ def __init__(self, trainer):
...
+
+    @return_predictions.setter
+    def return_predictions(self, return_predictions: Optional[bool] = None) -> None:
+        if return_predictions and self.trainer.training_type_plugin.use_spawn:
+            raise MisconfigurationException(
+                "`return_predictions` should be set to `False` when using spawn accelerators. "
+                f"Found {return_predictions}"
+            )
+        if return_predictions is not None:
+            self._return_predictions = return_predictions
```
Comment (awaelchli) [inline pytorch_lightning/trainer/predict_loop.py:48]: add a simple cpu test for this flag?

File: pytorch_lightning/trainer/predict_loop.py [hunk 30]
```diff
@@ -24,6 +27,24 @@ def __init__(self, trainer):
...
+        self._return_predictions = not trainer.training_type_plugin.use_spawn
+        self._previous_grad_status: Optional[bool] = None
+
+    @property
+    def return_predictions(self) -> bool:
+        return self._return_predictions
+
+    @return_predictions.setter
+    def return_predictions(self, return_predictions: Optional[bool] = None) -> None:
+        if return_predictions and self.trainer.training_type_plugin.use_spawn:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/predict_loop.py:42]:
```suggestion
        if return_predictions and isinstance(training_type_plugin, DDPSpawnPlugin):
```

File: pytorch_lightning/trainer/predict_loop.py [hunk 31]
```diff
@@ -24,6 +27,24 @@ def __init__(self, trainer):
...
+            raise MisconfigurationException(
+                "`return_predictions` should be set to `False` when using spawn accelerators. "
+                f"Found {return_predictions}"
```
Comment (awaelchli) [inline pytorch_lightning/trainer/predict_loop.py:45]:
```suggestion
                "`return_predictions` should be set to `False` when using the `DDPSpawnPlugin`."
                f" Found {return_predictions}"
```

**Review by awaelchli — COMMENTED** [review]
why do you make the returning of predictions optional? what difference does it make?

File: tests/overrides/test_distributed.py (hunk 19 above)
Comment (tchaton) [reply tests/overrides/test_distributed.py:27]: Done !

File: tests/overrides/test_distributed.py (hunk 22 above)
Comment (tchaton) [reply tests/overrides/test_distributed.py:49]: Done !

File: pytorch_lightning/plugins/training_type/parallel.py (hunk 24 above)
Comment (tchaton) [reply pytorch_lightning/plugins/training_type/parallel.py:52]: Yes, but I find it less clean.

File: pytorch_lightning/trainer/data_loading.py (hunk 25 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:154]: Yes, by default the trainer will make a prediction for all available samples in the dataset.

File: pytorch_lightning/trainer/predict_loop.py (hunk 26 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:147]: Done !

File: pytorch_lightning/trainer/predict_loop.py (hunk 27 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:143]: Done !

File: pytorch_lightning/trainer/trainer.py (hunk 28 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1045]: Done !

File: pytorch_lightning/trainer/predict_loop.py (hunk 30 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py:42]: Done !

File: pytorch_lightning/trainer/predict_loop.py (hunk 29 above)
Comment (tchaton) [reply pytorch_lightning/trainer/predict_loop.py]: Done !

File: CHANGELOG.md [hunk 32]
```diff
@@ -111,6 +111,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added new `EarlyStopping` parameters `stopping_threshold` and `divergence_threshold` ([#6868](https://github.com/PyTorchLightning/pytorch-lightning/pull/6868))
 
 
+- Added new `UnrepeatedDistributedSampler` and `IndexBatchSamplerWrapper` for tracking distributed predictions ([#7215](https://github.com/PyTorchLightning/pytorch-lightning/pull/7215))
+
+- Added `trainer.predict(return_predictions=None|False|True)` ([#7215](https://github.com/PyTorchLightning/pytorch-lightning/pull/7215))
```
Comment (carmocca) [inline CHANGELOG.md:116]:
```suggestion

- Added `trainer.predict(return_predictions=None|False|True)` ([#7215](https://github.com/PyTorchLightning/pytorch-lightning/pull/7215))
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/overrides/distributed.py (hunk 15 above)
Comment (SeanNaren) [inline pytorch_lightning/overrides/distributed.py:81]: Is `NoPaddingDistributedSampler` clearer?

**Review by SeanNaren — APPROVED** [review]
Nice cleanup as well @tchaton :)

**Comment by Borda** [comment]
> why do you make the returning of predictions optional? what difference does it make?

Any comment here? @tchaton

File: pytorch_lightning/plugins/training_type/parallel.py (hunk 24 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/parallel.py]: The problem is this style of adding properties that identify each plugin will not scale long term and at the finest granularity it is the same as an instance check. 

it could be part of the accelerator connector refactor one day. The use_spawn could be handled by the connector.
